internal/socket/socket.go

```diff
@@ -0,0 +1,163 @@
+//go:build windows
+
+package socket
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"golang.org/x/sys/windows"
+)
+
+//go:generate go run golang.org/x/sys/windows/mkwinsyscall -output zsyscall_windows.go socket.go
+
+//sys getsockname(s windows.Handle, name unsafe.Pointer, namelen *int32) (err error) [failretval==socketError] = ws2_32.getsockname
+//sys getpeername(s windows.Handle, name unsafe.Pointer, namelen *int32) (err error) [failretval==socketError] = ws2_32.getpeername
+//sys bind(s windows.Handle, name unsafe.Pointer, namelen int32) (err error) [failretval==socketError] = ws2_32.bind
+
+const socketError = uintptr(^uint32(0))
+
+var ErrSocketClosed = fmt.Errorf("socket closed: %w", net.ErrClosed)
+
+// GetSockName returns the socket's local address. It will call the `rsa.FromBytes()` on the
+// buffer returned by the getsockname syscall. The buffer is allocated to the size specified
+// by `rsa.Sockaddr()`.
+func GetSockName(s windows.Handle, rsa RawSockaddr) error {
+	ptr, l, err := rsa.Sockaddr()
+	if err != nil {
+		return fmt.Errorf("could not find socket size to allocate buffer: %w", err)
+	}
+	if err = validateSockAddr(ptr, l); err != nil {
+		return err
+	}
```

Comment:
I'd still like for the impls to verify inside

Reply:
Done, and since we assume the caller is returning a valid pointer, we can also get rid of `FromBytes` and overwrite the memory pointed to ourselves.